### ISSUE TITLE
How to allow a single version of a module to be compatible with both D7 and Backdrop 1.x?

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5024]
See:

- [Drupal compatibility layer](https://docs.backdropcms.org/api/backdrop/core%21includes%21drupal.inc/group/drupal_compatibility/1)
  > These functions and classes are wrappers around Backdrop functions that allow for Drupal 7 modules to work with Backdrop. All of these functions are made available if the "backdrop_drupal_compatibility" option is enabled in your site's settings.php file.
- [Converting modules to Backdrop from Drupal 7](https://docs.backdropcms.org/converting-modules-from-drupal)
  > The first and most important change is that a new Backdrop-specific core version needs to be added: `'backdrop = 1.x'`. For modules that would like to support both Drupal 7 and Backdrop at the same time, the old Drupal core version can also remain, but for Backdrop-only modules this line should be removed: `core = 7.x`
- Also, here's what's by default in our settings.php OOTB:
  
  ```php
  /**
   * Drupal backwards compatibility.
   *
   * By default, Backdrop 1.0 includes a compatibility layer to keep it compatible
   * with Drupal 7 APIs. Backdrop core itself does not use this compatibility
   * layer however. You may disable it if all the modules you're running were
   * built for Backdrop.
   */
  $settings['backdrop_drupal_compatibility'] = TRUE;
  ```

While commenting in https://github.com/backdrop/backdrop-issues/issues/5002#issuecomment-808674484 I tried the following on my local:

```
name = Metatag
description = "Adds support and an API to implement meta tags."
package = SEO
core = 7.x
backdrop = 1.x
```

Which resulted in this:

![image](https://user-images.githubusercontent.com/2423362/112714071-3309d400-8ee1-11eb-84f2-fb90b08d4e08.png)

The .info file for this module includes the following when it comes to dependencies:

```
; This requires Drupal 7.40 because of the project:module dependency structure.
dependencies[] = drupal:system (>= 7.40)

; CTools is required.
dependencies[] = ctools:ctools

; Requires Token; v7.x-1.6 is highly recommended due to bugs with certain tags
; when using older versions.
dependencies[] = token:token
```

Should we also allow prefixing dependencies with `backdrop:` like this then?:

```
dependencies[] = drupal:system (>= 7.40)
dependencies[] = backdrop:system (>= 1.19)
```

On our end, we could change the dependencies logic to translate the above to "`drupal:system (>= 7.40)` OR `backdrop:system (>= 1.19)`" (basically ignore any dependency that starts with `drupal:`), but it would stop the module from working in D7 (because of the `backdrop:` dependency) 🤔 

And how could we "ignore" certain packages in the Backdrop-specific dependencies? What would be the respective line so that these are not a dependency issue?:

```
dependencies[] = ctools:ctools
dependencies[] = token:token
```

**Comment by indigoxela** [comment]
Is this a theoretical question, or does any module exist, that really tries to be compatible with both, Drupal and Backdrop?

Your example (metatag) got fully ported. Personally, I'd never try to maintain something complex like that. Code that relies on ctools when run on D7, but does something totally different when run on Backdrop. :smile:

**Comment by klonos** [comment]
> Is this a theoretical question...

Yup, theoretical. In theory, any "simple" D7 module that doesn't have any settings (hence no need to be adapted to save in config instead of the db), should be able to work with both D7 and Backdrop. Our documentation implies that that is the case (see references in the issue summary).

**Comment by indigoxela** [comment]
> The question is how do we make Backdrop ignore the first dependency, and Drupal ignore the second?

We could - in theory - implement checks for CMS related dependencies in Backdrop. We can not in Drupal. That's why this will probably stay a theoretical issue. :wink: 

Module maintainers can do those checks in an abstraction layer in their code. And there's also hook_requirements(). But who would voluntarily do something complicated like that? IMO it's much easier to maintain two separate projects then.

There *are* modules that could work for both, B and D7 - modules that only use hooks that didn't change in B. But these modules are very easy to port, either.

**Comment by klonos** [comment]
> IMO it's much easier to maintain two separate projects then.

Agreed. Although if the only difference in two separate branches of the same project is a single line in their .info file, and no other change in the actual code (assuming `$settings['backdrop_drupal_compatibility'] = TRUE;`), then it seems like a pointless overhead to be managing 2 separate branches. Right?
